### ISSUE TITLE
[Bug?]: Peer dependencies not met for some @types

### ISSUE THREAD

**Issue by donferi** [yarnpkg/berry#5195]
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

We are getting an incorrect dependency met warning with `@types/react-datepicker`, which has a dependency on `react-popper` which has a `peerDependency` on react.

Since `@types/*` can't have `peerDependencies` I'm not sure how this should be handled. In a regular scenario `@types/react-datepicker` would define a `peerDependency` on `react` and the root project's dependency on react would satisfies this. But unsure how it should be done in this case.

### To reproduce

```js repro
await packageJsonAndInstall({
  dependencies: {
    [`@popperjs/core`]: `^2.11.6`,
    [`@types/react-datepicker`]: `^4.8.0`,
    [`react`]: `^18.2.0`,
    [`react-dom`]: `^18.2.0`,
    [`react-popper`]: `^2.3.0`,
    [`typescript`]: `^4.9.4`
  }
});

await expect(yarn(`install`)).resolves.not.toContain(`peer dependencies are incorrectly met`);
```

### Environment

```shell
System:
    OS: macOS 13.1
    CPU: (10) x64 Apple M1 Pro
  Binaries:
    Node: 14.19.3 - /private/var/folders/l2/ym3fhv6n28l6bwt_mjglk6ww0000gq/T/xfs-243c2d13/node
    Yarn: 4.0.0-rc.35.git.20230109.hash-c1919d3 - /private/var/folders/l2/ym3fhv6n28l6bwt_mjglk6ww0000gq/T/xfs-243c2d13/yarn
    npm: 6.14.17 - ~/Library/Caches/fnm_multishells/14357_1673541529681/bin/npm
```

**Comment by yarnbot** [comment]
This issue reproduces on master:
```
Error: expect(received).resolves.not.toContain(expected) // indexOf

Expected substring: not "peer dependencies are incorrectly met"
Received string:        "➤ YN0000: ┌ Resolution step
::group::Resolution step
➤ YN0002: │ @types/react-datepicker@npm:4.8.0 doesn't provide react (p404f9), requested by react-popper
➤ YN0002: │ @types/react-datepicker@npm:4.8.0 doesn't provide react-dom (p2a659), requested by react-popper
➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code
::endgroup::
➤ YN0000: └ Completed
➤ YN0000: ┌ Fetch step
::group::Fetch step
➤ YN0013: │ 17 packages were already cached
::endgroup::
➤ YN0000: └ Completed
➤ YN0000: ┌ Link step
::group::Link step
::endgroup::
➤ YN0000: └ Completed
➤ YN0000: Done with warnings in 0s 281ms
"
    at Object.toContain (/github/workspace/.yarn/cache/expect-npm-24.8.0-8c7640c562-44ff9ab1e7.zip/node_modules/expect/build/index.js:202:20)
    at module.exports (evalmachine.<anonymous>:13:44)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async /github/workspace/.yarn/cache/@arcanis-sherlock-npm-2.0.3-558f52b79f-286d94b96d.zip/node_modules/@arcanis/sherlock/lib/executeRepro.js:57:13
    at async executeInTempDirectory (/github/workspace/.yarn/cache/@arcanis-sherlock-npm-2.0.3-558f52b79f-286d94b96d.zip/node_modules/@arcanis/sherlock/lib/executeRepro.js:18:16)
    at async executeRepro (/github/workspace/.yarn/cache/@arcanis-sherlock-npm-2.0.3-558f52b79f-286d94b96d.zip/node_modules/@arcanis/sherlock/lib/executeRepro.js:25:12)
    at async ExecCommand.execute (/github/workspace/.yarn/cache/@arcanis-sherlock-npm-2.0.3-558f52b79f-286d94b96d.zip/node_modules/@arcanis/sherlock/lib/commands/exec.js:26:38)
    at async ExecCommand.validateAndExecute (/github/workspace/.yarn/cache/clipanion-npm-2.0.0-rc.16-b9444aaf89-4061026d74.zip/node_modules/clipanion/lib/advanced/Command.js:161:26)
    at async Cli.run (/github/workspace/.yarn/cache/clipanion-npm-2.0.0-rc.16-b9444aaf89-4061026d74.zip/node_modules/clipanion/lib/advanced/Cli.js:74:24)
    at async Cli.runExit (/github/workspace/.yarn/cache/clipanion-npm-2.0.0-rc.16-b9444aaf89-4061026d74.zip/node_modules/clipanion/lib/advanced/Cli.js:83:28)
```

**Comment by arcanis** [comment]
It's a little tricky; the right solution would be to use peer dependencies, as you said, but if DefinitelyTyped is hard-bent on not using them (which is very unfortunate, they exist for a reason), your best shot is to use [`packageExtensions`](https://yarnpkg.com/configuration/yarnrc/#packageExtensions) to define the missing peer dependency yourself.

**Comment by donferi** [comment]
Indeed it's unfortunate, my first react was to open a PR to `DefinitelyTyped` but running the tests locally failed because of it including peer deps (unless I somehow missed something). 

I went ahead and used your `packageExtensions` recommendation and worked beautifully.

I still wonder how other package manager solve this, is it because `pnpm` for example is not as strict with peer deps? 🤔  Just tried with it just to see if I got the same warning but didn't.

PS: Closing as this is not actionable.
